### PR TITLE
Deciding between Cloud Spanner and Microsoft Cosmos DB 

### LINKED ISSUE #542: Deciding between Cloud Spanner and Microsoft CosmosDB
### Brief Summary: 
Describe the what, why, and how of your content idea in 2-5 sentences.
With the advancements that are already happening within cloud computing, tons of improvements are bound to happen in the future. One such development includes the initiation of SQL based databases that provide the users with a customized experience of storing, sorting, and processing heavy bounds of raw data. There are many database providers out there and most of them work on various nodes of the SQL database engines such as Google Spanner and Microsoft CosmosDB. Deciding on which cloud database to use is usually an uphill task. 
### Key Takeaways: 
What are the 3-5 most important things the reader should understand or be able to do after reading this article?
1. Salient features of Cloud Spanner and Microsoft CosmosDB 
2. Merits and Demerits of Cloud Spanner and Microsoft CosmosDB
3. Differences between Cloud Spanner and Microsoft CosmosDB
### References: 
Please list links to any published content/research that you intend to use to support/guide this article. (If none, please indicate 'N/A'.)
N/A

### PR DESCRIPTION
**Pull request by EKahuha**
This fixes #542

#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      

#### Grammar Checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical Checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution Guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**

### CODE REVIEW

**Comment by hectorkambow** [comment]
👋 Thank you for submitting this PR - we will review it ASAP - all PRs go into our queue will be looked over before being approved to be published.

**Comment by hectorkambow** [comment]
plagiarism check done

File: articles/deciding-between-cloud-spanner-and-microsoft-cosmos-db/index.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
+Creating a globally distributed database would traditionally require a lot of time and effort. You had to host the database in an own data center and use your own connections. Today&#39;s advancement in cloud computing and platforms as a service (PaaS) has made it easier to create globally distributed and scalable databases. Particularly, the use of SQL-based databases has become more popular in the recent past. These databases provide you with a customized experience of sorting and processing large amounts of raw data. Many database providers exist, and they work on different nodes of the SQL database engines, including [Microsoft Azure Cosmos DB](https://docs.microsoft.com/en-us/azure/cosmos-db/introduction) and [Google Cloud Spanner](https://cloud.google.com/spanner).
```
Comment (lalith1403) [inline articles/deciding-between-cloud-spanner-and-microsoft-cosmos-db/index.md:1]: You would ou had to host the database in an own data center and use your own connections->have to host the database in a personal data center bearing the entire cost.

File: articles/deciding-between-cloud-spanner-and-microsoft-cosmos-db/index.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+This article will discuss the features of both Cloud Spanner and Microsoft Cosmos DB and highlight their advantages and disadvantages. It will also compare the two to help you better choose the cloud database to use in the future.
+
+### Features of Google Cloud Spanner
+
+Google Spanner is a fully managed, scalable, relational database management service.
+
+The platform can scale millions of nodes across multiple regions. Besides being integrated with access and identity management, it provides features such as [logging and auditing](https://cloud.google.com/spanner/docs/audit-logging).
```
Comment (lalith1403) [inline articles/deciding-between-cloud-spanner-and-microsoft-cosmos-db/index.md:9]: Can you define what a node is and define its features?

File: articles/deciding-between-cloud-spanner-and-microsoft-cosmos-db/index.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+Google Cloud Spanner&#39;s instance is comparable to a relational database management system ([RDBMS](https://www.codecademy.com/articles/what-is-rdbms-sql)) server. This instance is composed of one or multiple databases, which use the same [compute](https://www.zdnet.com/article/what-is-cloud-computing-everything-you-need-to-know-about-the-cloud/) and [storage](https://searchstorage.techtarget.com/definition/cloud-storage) resources. These resources are allocated during [instance](https://cloud.google.com/spanner/docs/instances) creation. Two main configurations, [Regional and Multi-Regional](https://cloud.google.com/spanner/docs/instances), control the compute and storage resources.
```
Comment (lalith1403) [inline articles/deciding-between-cloud-spanner-and-microsoft-cosmos-db/index.md:11]: The last sentence is not clear. 2 main configurations are regional and multi-regional. But how do they control the compute and storage resources? Some more emphasis on how distributed databases come into cost computation would be better.

File: articles/deciding-between-cloud-spanner-and-microsoft-cosmos-db/index.md [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+Cloud Spanner supports automatic data replication. Instance configuration (the first configuration) determines the number of copies ([replicas](https://cloud.google.com/spanner/docs/replication)) to be created and their placement. Regional configuration allows data replication across three zones. These zones must come from within a single selected region. Multi-Regional configuration, on the other hand, supports data replication across four zones. These zones can come from different regions depending on what continent the user specifies.
```
Comment (lalith1403) [inline articles/deciding-between-cloud-spanner-and-microsoft-cosmos-db/index.md:13]: Include a map of GCP zones and regions. It's not clear what these conditions are, and how they are useful. A picture should solve that problem.

File: articles/deciding-between-cloud-spanner-and-microsoft-cosmos-db/index.md [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+### Features of Cosmos DB
+
+Microsoft Cosmos DB is a Platform as a Service ([PaaS](https://searchcloudcomputing.techtarget.com/definition/Platform-as-a-Service-PaaS)) service in [Microsoft Azure](https://azure.microsoft.com/en-us/). PaaS is a deployment environment in the cloud that provides a platform for individuals and organizations to develop, manage, and run cloud-enabled enterprise applications. Azure refers to Microsoft&#39;s public cloud offering. With Azure Cosmos DB, users can build and distribute their applications across an Azure data center. Manual work and configuration is eliminated, in this case.
+
+The globally distributed, multi-model database is available across more than 30 regions worldwide. This is possible considering its turnkey global distribution that automatically scales and replicates data across various data centers in the Azure network.
+
+Cosmos DB&#39;s multi-model feature supports multiple models, including table storage, graphs, key values, and document storage in a single database. It offers high availability and consistency across all these data models, irrespective of their nature, for data storage.
```
Comment (lalith1403) [inline articles/deciding-between-cloud-spanner-and-microsoft-cosmos-db/index.md:25]: key-values

File: articles/deciding-between-cloud-spanner-and-microsoft-cosmos-db/index.md [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+
+### Features of Cosmos DB
+
+Microsoft Cosmos DB is a Platform as a Service ([PaaS](https://searchcloudcomputing.techtarget.com/definition/Platform-as-a-Service-PaaS)) service in [Microsoft Azure](https://azure.microsoft.com/en-us/). PaaS is a deployment environment in the cloud that provides a platform for individuals and organizations to develop, manage, and run cloud-enabled enterprise applications. Azure refers to Microsoft&#39;s public cloud offering. With Azure Cosmos DB, users can build and distribute their applications across an Azure data center. Manual work and configuration is eliminated, in this case.
+
+The globally distributed, multi-model database is available across more than 30 regions worldwide. This is possible considering its turnkey global distribution that automatically scales and replicates data across various data centers in the Azure network.
```
Comment (lalith1403) [inline articles/deciding-between-cloud-spanner-and-microsoft-cosmos-db/index.md:23]: A map would be useful

File: articles/deciding-between-cloud-spanner-and-microsoft-cosmos-db/index.md [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+On the flip side, users associate Cosmos DB platform with some imperfections. Cosmos DB fails to provide a seamless way to switch between [database-provisioned throughput and container-provisioned throughput](https://docs.microsoft.com/en-us/azure/cosmos-db/set-throughput). You have to recreate the database for you to switch successfully. Besides, Cosmos DB does not allow documents from different logical partition to participate in the same transaction. And writes from different containers cannot be part of the same transaction.
```
Comment (lalith1403) [inline articles/deciding-between-cloud-spanner-and-microsoft-cosmos-db/index.md:35]: What does this mean? Database-provisioned throughput and container-provisioned throughput. Could you explain that or attach a link to that. The link you have given does not give information about container provisional throughput.

File: articles/deciding-between-cloud-spanner-and-microsoft-cosmos-db/index.md [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+### Pros and Cons of Google Cloud Spanner
+
+One of the strengths of Google Cloud Spanner is its strong offering in containers. The [Kubernetes standard](https://cloud.google.com/kubernetes-engine) developed by Google is now also offered by Azure. Google Cloud Platform ([GCP](https://console.cloud.google.com/)) deals with high compute offerings, such as big data, machine learning, and analytics. GCP provides significant scale and load balancing.
+
+Spanner is capable of handling large volumes of data. Its use is not limited to applications of large sizes. Further, it allows the standardization of a single database engine for all workloads requiring an RDBMS. This is very beneficial to organizations.
+
+On the downside, it may be challenging to create an instance in a local environment for Cloud Spanner. A development environment needs to be as close a match to production as possible. But this is not the case with Spanner because you must rely on a full spanner instance to accomplish this need. You may choose a single region instance to save on costs.
```
Comment (lalith1403) [inline articles/deciding-between-cloud-spanner-and-microsoft-cosmos-db/index.md:43]: A development environment needs to be as close to production as possible.

File: articles/deciding-between-cloud-spanner-and-microsoft-cosmos-db/index.md [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+
+### Major Differences between Cloud Spanner and Microsoft Cosmos DB
+
+Google Spanner supports concurrent ACID (atomicity, consistency, isolation, and durability) guarantees across more than one operation. Cosmos DB supports ACID only within the limit of a single operation. ACID are properties that ensure reliable processing of [database transactions](https://en.wikipedia.org/wiki/Database_transaction). ACID databases guarantee data correctness and maintainability in extensive production systems.
```
Comment (lalith1403) [inline articles/deciding-between-cloud-spanner-and-microsoft-cosmos-db/index.md:47]: There must be a link on ACID operations in the eng ed community. Link it here if you find it.

**Review by lalith1403 — CHANGES_REQUESTED** [review]
Well written article. I have included a few points where the additional explanation would help the reader. The last bit where you discuss the pros and cons of the 2 services is well written.

**Comment by hectorkambow** [comment]
plagiarism check done

File: articles/deciding-between-cloud-spanner-and-microsoft-cosmos-db/index.md (hunk 7 above)
Comment (EKahuha) [reply articles/deciding-between-cloud-spanner-and-microsoft-cosmos-db/index.md:35]: > What does this mean? Database-provisioned throughput and container-provisioned throughput. Could you explain that or attach a link to that. The link you have given does not give information about container provisional throughput.

@lalith1403 Container-provisioned throughput is explained under the link provided. It is under the sub-heading "Set throughput on a container"

Comment (lalith1403) [reply articles/deciding-between-cloud-spanner-and-microsoft-cosmos-db/index.md:35]: You may guide the reader to that sub section. The thing is, a reader has limited attention. Including several links leads to dispersion in attention and therefore adds less value.

File: articles/deciding-between-cloud-spanner-and-microsoft-cosmos-db/index.md (hunk 9 above)
Comment (EKahuha) [reply articles/deciding-between-cloud-spanner-and-microsoft-cosmos-db/index.md:47]: > There must be a link on ACID operations in the eng ed community. Link it here if you find it.

I could not find an article on ACID operations in the eng-ed community. I think it should be my next topic.

**Comment by EKahuha** [comment]
@lalith1403 
I have updated the file.

**Review by lalith1403 — APPROVED** [review]

**Pull request merged**